### PR TITLE
[Offline Mode] - Add support for caching categories

### PR DESCRIPTION
**Pull request by AnotherMedo**
The last piece of the puzzle that is this _**offline mode**_ is to cache the categories to allow to consult them offline (and also because it doesn't make much sense to get them from network every time as they are _largely_ static)

**Commit history:**
`task: add main backend for category caching`
- [x] Add an adequate `DAO` for categories
- [x] Add `CategoryEntity` in order to serialize categories
- [x] Modify the `CategoryRepositoryFirestore` to query from cache when possible
- [x] Modify the `CategoryViewModel` to reflect changes to the repository
- [x] Add proper serializing and de-serializing functionality for categories and subcategories in `Converters`
- [x] Make additional tweaks and changes

`test: write and update tests to include caching`
- [x] Add tests for all added classes
- [x] Update other tests to reflect changes done 

`task: update files to new spec`
- [x] Update all necessary files for compilation errors

### CODE REVIEW

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepository.kt [hunk 1]
```diff
@@ -4,15 +4,15 @@ interface CategoryRepository {
 
   fun init(onSuccess: () -> Unit)
 
-  fun fetchCategories(onSuccess: (List<Category?>) -> Unit, onFailure: (Exception) -> Unit)
+  suspend fun fetchCategories(onSuccess: (List<Category?>) -> Unit, onFailure: (Exception) -> Unit)
```
Comment (Sebbayra) [inline app/src/main/java/com/arygm/quickfix/model/category/CategoryRepository.kt:7]: Can you explain what a suspend fun is?

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt [hunk 2]
```diff
@@ -23,60 +27,83 @@ open class CategoryRepositoryFirestore(private val db: FirebaseFirestore) : Cate
...
-        withContext(Dispatchers.Main) { onSuccess(categories) }
-      } catch (e: Exception) {
-        Log.e("fetchCategories", "Error fetching categories", e)
-        withContext(Dispatchers.Main) { onFailure(e) }
+    try {
+      // Attempt to fetch from Room first
+      val cachedCategories =
+          dao.getAllCategories()
+              .map { entities -> entities.map { it.toCategory() } }
+              .first() // Take only the first emission from the flow
```
Comment (Sebbayra) [inline app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt:39]: Really like this, seeing as a flow returns it continuously, we just take the first output, gjgjgj

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt [hunk 3]
```diff
@@ -91,15 +118,37 @@ open class CategoryRepositoryFirestore(private val db: FirebaseFirestore) : Cate
...
+      }
+
+      Log.d("fetchSubcategoriesSuspend", "Returning Firestore subcategories for $categoryId")
+      subcategories
     } catch (e: Exception) {
-      Log.e("fetchSubcategoriesSuspend", "Error fetching documents from Firestore", e)
-      null
+      Log.e("fetchSubcategoriesSuspend", "Error fetching subcategories for $categoryId", e)
+      emptyList() // Return an empty list in case of an error
     }
```
Comment (Sebbayra) [inline app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt:152]: I'm quite confused here, the function does the same thing as before but isn't asynchronous can we just use the one from before but with a correct onSuccess?

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryViewModel.kt
Comment (Sebbayra) [inline app/src/main/java/com/arygm/quickfix/model/category/CategoryViewModel.kt:1]: Clean, as always goated activities

**Review by Sebbayra — APPROVED** [review]
Great job, clean implementation, only thing to open an issue about is the redundant fuction.

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepository.kt (hunk 1 above)
Comment (AnotherMedo) [reply app/src/main/java/com/arygm/quickfix/model/category/CategoryRepository.kt:7]: A `suspend` function is a function that can be paused and resumed later. It is designed to work with **coroutines**, which are a powerful tool for handling asynchronous or long-running operations in Kotlin.

When a `suspend` function is called, it doesn't block the calling thread. Instead, it suspends execution and allows other code to run until the function resumes. 

By all means don't hesitate to ask if anything else needs clarification 😄

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt (hunk 2 above)
Comment (AnotherMedo) [reply app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt:39]: 🐐 🔥

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt (hunk 3 above)
Comment (AnotherMedo) [reply app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt:152]: It is not me that implemented this function but an issue could definitely be opened for this 😉

File: app/src/main/java/com/arygm/quickfix/model/category/CategoryViewModel.kt
Comment (AnotherMedo) [reply app/src/main/java/com/arygm/quickfix/model/category/CategoryViewModel.kt:1]: thank 🚀

**Pull request merged**